### PR TITLE
[core] Freenoise memory improvements

### PR DESCRIPTION
**Pull request by a-r-r-o-w**
# What does this PR do?

Memory improvements from #9231. The previous PR had too many changes so it has been split to make it easier to review. This PR contains just the FreeNoise memory improvements whereas the previous contains prompt travel support. This PR will be ready for review after 9231 has been merged and this has been rebased.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@DN6

### CODE REVIEW

File: src/diffusers/models/attention_processor.py [hunk 1]
```diff
@@ -2372,6 +2372,8 @@ def __call__(
         hidden_states = hidden_states.to(query.dtype)
 
         # linear proj
+        # TODO: figure out a better way to do this
+        # hidden_states = torch.cat([attn.to_out[1](attn.to_out[0](x)) for x in hidden_states.split(4, dim=0)], dim=0)
```
Comment (DN6) [inline src/diffusers/models/attention_processor.py:2376]: Will this line be applied? If not, we can remove.

File: src/diffusers/pipelines/free_noise_utils.py [hunk 2]
```diff
@@ -410,6 +467,54 @@ def disable_free_noise(self) -> None:
...
+                )
+
+            motion_module.proj_out = SplitInferenceModule(motion_module.proj_out, spatial_split_size, 0, ["input"])
+
+    def _enable_split_inference_attentions_(
+        self, attentions: List[Transformer2DModel], temporal_split_size: int
+    ) -> None:
+        for i in range(len(attentions)):
+            attentions[i] = SplitInferenceModule(
+                attentions[i], temporal_split_size, 0, ["hidden_states", "encoder_hidden_states"]
```
Comment (sayakpaul) [inline src/diffusers/pipelines/free_noise_utils.py:491]: Should `["hidden_states", "encoder_hidden_states"]` not be configurable or not really?

File: src/diffusers/pipelines/free_noise_utils.py [hunk 3]
```diff
@@ -410,6 +467,54 @@ def disable_free_noise(self) -> None:
...
+                self._enable_split_inference_motion_modules_(block.motion_modules, spatial_split_size)
+            if getattr(block, "attentions", None) is not None:
+                self._enable_split_inference_attentions_(block.attentions, temporal_split_size)
+            if getattr(block, "resnets", None) is not None:
+                self._enable_split_inference_resnets_(block.resnets, temporal_split_size)
+            if getattr(block, "downsamplers", None) is not None:
+                self._enable_split_inference_samplers_(block.downsamplers, temporal_split_size)
+            if getattr(block, "upsamplers", None) is not None:
+                self._enable_split_inference_samplers_(block.upsamplers, temporal_split_size)
+
```
Comment (sayakpaul) [inline src/diffusers/pipelines/free_noise_utils.py:517]: Same. Should `attentions`, `resnets`, etc. ne not configurable?

File: src/diffusers/pipelines/free_noise_utils.py [hunk 4]
```diff
@@ -30,6 +34,53 @@
...
+        self.module = module
+        self.split_size = split_size
+        self.split_dim = split_dim
+        self.input_kwargs_to_split = set(input_kwargs_to_split)
+
+    def forward(self, *args, **kwargs) -> Union[torch.Tensor, Tuple[torch.Tensor]]:
+        r"""Forward method of `SplitInferenceModule`.
+
+        All inputs that should be split should be passed as keyword arguments. Only those keywords arguments will be
+        split that are specified in `inputs_to_split` when initializing the module.
```
Comment (sayakpaul) [inline src/diffusers/pipelines/free_noise_utils.py:56]: Perhaps this could be expanded to include a little summary of what's being done?

File: src/diffusers/models/unets/unet_motion_model.py [hunk 5]
```diff
@@ -1080,10 +1066,19 @@ def forward(
...
         blocks = zip(self.attentions, self.resnets[1:], self.motion_modules)
         for attn, resnet, motion_module in blocks:
+            hidden_states = attn(
+                hidden_states=hidden_states,
+                encoder_hidden_states=encoder_hidden_states,
+                cross_attention_kwargs=cross_attention_kwargs,
+                attention_mask=attention_mask,
+                encoder_attention_mask=encoder_attention_mask,
+                return_dict=False,
+            )[0]
```
Comment (sayakpaul) [inline src/diffusers/models/unets/unet_motion_model.py:1080]: Are we not allowing gradient checkpointing or maybe I am misunderstanding the change?

File: src/diffusers/models/attention.py [hunk 6]
```diff
@@ -1104,8 +1104,15 @@ def forward(
...
-            num_times_accumulated > 0, accumulated_values / num_times_accumulated, accumulated_values
+        hidden_states = torch.cat(
+            [
+                torch.where(num_times_split > 0, accumulated_split / num_times_split, accumulated_split)
+                for accumulated_split, num_times_split in zip(
+                    accumulated_values.split(self.context_length, dim=1),
+                    num_times_accumulated.split(self.context_length, dim=1),
+                )
+            ],
+            dim=1,
```
Comment (sayakpaul) [inline src/diffusers/models/attention.py:1115]: So, this seems to be a form of chunking?

**Review by sayakpaul — COMMENTED** [review]
I was not requested for a review but still did to learn some things. Thanks, Aryan!

File: src/diffusers/models/attention.py (hunk 6 above)
Comment (a-r-r-o-w) [reply src/diffusers/models/attention.py:1115]: Yep. At some point lowering the peaks on memory traces, torch.where became the bottleneck. This was actually first noticed by @DN6 so credits to him

File: src/diffusers/models/unets/unet_motion_model.py (hunk 5 above)
Comment (a-r-r-o-w) [reply src/diffusers/models/unets/unet_motion_model.py:1080]: There was no gradient checkpointing here to begin with actually. Both if-branches had the same code so I moved them out. If we'd like to add gradient checkpointing support, let's do it in a separate PR?

File: src/diffusers/pipelines/free_noise_utils.py (hunk 4 above)
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/free_noise_utils.py:56]: Sounds good. It got way too complicated way too fast when I was just trying to get it to work with all the zip and `*` expand

File: src/diffusers/pipelines/free_noise_utils.py (hunk 2 above)
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/free_noise_utils.py:491]: Sorry, I don't understand the comment very well.

`SplitInferenceModule` sets `input_kwargs_to_split` to `["hidden_states"]` by default if no parameter is passed. I want both hidden_states and encoder_hidden_states to be split based on `split_size` here

File: src/diffusers/pipelines/free_noise_utils.py (hunk 3 above)
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/free_noise_utils.py:517]: Not sure I understand this comment either. Basically, we're going to be splitting across the batch dimension for the layers based on chosen `spatial_split_size` and `temporal_split_size` values

File: src/diffusers/pipelines/free_noise_utils.py (hunk 2 above)
Comment (sayakpaul) [reply src/diffusers/pipelines/free_noise_utils.py:491]: Sorry. Okay. I am assuming that is a reasonable default to choose? I was wondering if it could make sense to let the users choose the inputs they wanna split?

File: src/diffusers/pipelines/free_noise_utils.py (hunk 3 above)
Comment (sayakpaul) [reply src/diffusers/pipelines/free_noise_utils.py:517]: Similarly. 

> Basically, we're going to be splitting across the batch dimension for the layers based on chosen spatial_split_size and temporal_split_size values

Could it make sense to let the users to choose the kind of layers that wanna apply splitting? Perhaps we default to all (`attentions`, `motion_modules`, `resnets`, `downsamplers`, `upsamplers`,) or not really?

File: src/diffusers/models/unets/unet_motion_model.py (hunk 5 above)
Comment (sayakpaul) [reply src/diffusers/models/unets/unet_motion_model.py:1080]: Ah okay. Makes perfect sense.

File: src/diffusers/models/attention.py (hunk 6 above)
Comment (sayakpaul) [reply src/diffusers/models/attention.py:1115]: Hmm, do we know the situations where `torch.where()` leads to spikes? Seems a little weird to me honestly.

File: src/diffusers/pipelines/free_noise_utils.py (hunk 2 above)
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/free_noise_utils.py:491]: I would say let's keep it in mind to allow users to have more control on this, but for now let's keep the scope of changes minimal. I would like to experiment on FreeNoise for CogVideoX as discussed internally, and so would like to get this in soon :)

Comment (sayakpaul) [reply src/diffusers/pipelines/free_noise_utils.py:491]: Okay then we could add a comment before the blocks that could be configured and revisit those if needed?

File: src/diffusers/models/attention.py (hunk 6 above)
Comment (a-r-r-o-w) [reply src/diffusers/models/attention.py:1115]: I think the spike that we see is due to tensors being copied. The intermediate dimensions for attention get large when generating many frames (let's say, 200+) here. We could do something different here too - I just did what seemed like the easiest thing to do (as these changes were made when I was trying out different things in quick succession to golf the memory spikes)

**Comment by a-r-r-o-w** [comment]
@sayakpaul WDYT about the explanation of SplitInferenceModule now?

File: src/diffusers/pipelines/free_noise_utils.py [hunk 7]
```diff
@@ -70,6 +168,9 @@ def _enable_free_noise_in_block(self, block: Union[CrossAttnDownBlockMotion, Dow
                     motion_module.transformer_blocks[i].load_state_dict(
                         basic_transfomer_block.state_dict(), strict=True
                     )
+                    motion_module.transformer_blocks[i].set_chunk_feed_forward(
```
Comment (DN6) [inline src/diffusers/pipelines/free_noise_utils.py:170]: Do we always need chunked feed forward set when enabling free noise? Might be overkill no?

File: src/diffusers/models/attention.py (hunk 6 above)
Comment (sayakpaul) [reply src/diffusers/models/attention.py:1115]: Ah cool. Let's perhaps make a note of this to reivsit later? At least this way, we are aware?

File: src/diffusers/pipelines/free_noise_utils.py [hunk 8]
```diff
@@ -30,6 +34,100 @@
...
+            >>> import torch
+
+            >>> model = nn.Linear(1000, 1000)
+            >>> split_module = SplitInferenceModule(
+            ...     model, split_size=2, split_dim=0, input_kwargs_to_split=["input_data"]
+            ... )
+
+            >>> input_tensor = torch.randn(42, 1000)
+            >>> # Will split the tensor into 21 slices of shape [2, 1000].
+            >>> output = split_module(input_data=input_tensor)
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/free_noise_utils.py:94]:
```suggestion
            >>> output = split_module(input=input_tensor)
```

My bad here, nn.Linear takes `input` here and not `input_data`

File: src/diffusers/pipelines/free_noise_utils.py [hunk 9]
```diff
@@ -30,6 +34,100 @@
...
+class SplitInferenceModule(nn.Module):
+    def __init__(
+        self,
+        module: nn.Module,
+        split_size: int = 1,
+        split_dim: int = 0,
+        input_kwargs_to_split: List[str] = ["hidden_states"],
+    ) -> None:
+        super().__init__()
+
```
Comment (DN6) [inline src/diffusers/pipelines/free_noise_utils.py:46]: Would also add a docstring here to explain the init arguments. Maybe the workflow example in forward can be moved up here too?

File: src/diffusers/pipelines/free_noise_utils.py [hunk 10]
```diff
@@ -30,6 +34,100 @@
...
+            that were passed.
+            3. The output tensors from each split are concatenated back together along `split_dim` before returning.
+
+        Example:
+            ```python
+            >>> import torch
+
+            >>> model = nn.Linear(1000, 1000)
+            >>> split_module = SplitInferenceModule(
+            ...     model, split_size=2, split_dim=0, input_kwargs_to_split=["input_data"]
```
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/free_noise_utils.py:89]:
```suggestion
            ...     model, split_size=2, split_dim=0, input_kwargs_to_split=["input"]
```

Same reason

File: src/diffusers/pipelines/free_noise_utils.py [hunk 11]
```diff
@@ -30,6 +34,100 @@
...
+            if key not in self.input_kwargs_to_split or not torch.is_tensor(kwargs[key]):
+                continue
+            split_inputs[key] = torch.split(kwargs[key], self.split_size, self.split_dim)
+            kwargs.pop(key)
+
+        # 2. Invoke forward pass across each split
+        results = []
+        for split_input in zip(*split_inputs.values()):
+            inputs = dict(zip(split_inputs.keys(), split_input))
+            inputs.update(kwargs)
```
Comment (DN6) [inline src/diffusers/pipelines/free_noise_utils.py:114]: Clean 😎

File: src/diffusers/pipelines/free_noise_utils.py (hunk 2 above)
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/free_noise_utils.py:537]: Thanks, updated. WDYT?

**Review by DN6 — APPROVED** [review]
PR looks really good 👍🏽  Could you just add a fast GPU test to verify that the split inference outputs and normal outputs are the same?

File: src/diffusers/pipelines/free_noise_utils.py (hunk 7 above)
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/free_noise_utils.py:170]: This is only there to carry forward the chunk FF behaviour if and only if it was already enabled in the BasicTransformerBlock. Basically, if it was not enable in BTB, `motion_module.transformer_blocks[i]._chunk_size` would be None leading to default behaviour of no chunking. If it was enabled in BTB, it would by default carry forward to FreeNoiseTransformerBlock

File: src/diffusers/pipelines/free_noise_utils.py (hunk 9 above)
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/free_noise_utils.py:46]: Sounds good!

File: src/diffusers/pipelines/free_noise_utils.py (hunk 11 above)
Comment (a-r-r-o-w) [reply src/diffusers/pipelines/free_noise_utils.py:114]: Not too proud of this one :grimacing: but hey, I'm not afraid of not understanding what sorcery is going on here two weeks later

image

File: src/diffusers/models/attention.py (hunk 6 above)
Comment (a-r-r-o-w) [reply src/diffusers/models/attention.py:1126]: Alright, made a note. LMK if any further changes needed on this :)
